Issue8 -- map variadic

ericpoe opened the pull request:
This adds a variadic option to `HArray->map()` and `HString->map()` to fix Issue #8.

The variadic option keeps `->map()` closer in-line with `array_map` functionality.

ericpoe committed:
Add "toArray" method to the ContainerInterface

ericpoe committed:
Add variadic parameter to ->map()

This mimics the array_map($callback, $array, ...$otherArrays) in PHP5.6+

johnkary commented on tests/Functional/HStringMapTest.php:
```diff
@@ -39,4 +40,24 @@ public function testStringMapRot13()
+        };
+
+        $pairs = $this->aString->map($oddPairing, $myArray);
+        $expected = "-f:apple--o:butter--o:cookie--b:donut--a:eclaire--r:framboise-";
```
When implementing [array_zip()](https://github.com/johnkary/rethinkingloops/blob/master/examples/array_zip.php#L48) I had to account for cases where the two arrays differed in length. This includes:
1. $this->aString length shorter than $myArray
2. $this->aString length same length as $myArray
3. $this->aString length longer than $myArray

Perhaps see how array_map() works in those cases and mimic it?

ericpoe replied:
Good call. I was certain that this behaved the same way, but without tests, I cannot prove it. :)

ericpoe replied:
I've [added tests for HArray](9/commits/01b12295bf76d12f2b17c0ebac3a053374329d39). I'll do the same for HString later this evening.

ericpoe commented:
I need to figure out why HHVM doesn't like this. 7 errors along the lines of:

> Haystack\Tests\Functional\HArrayMapTest::testArrayMap
> Only containers may be unpacked

johnkary commented on src/HString.php:
```diff
@@ -232,10 +232,25 @@ public function toString()
     }
 
     /**
+     * Converts a string into an array. Assumes a delimiter of " ".
```
Does this method still assume delimiter of `" "` ? I think it was changed to return an array of characters.

ericpoe replied:
Good catch. Updated.

johnkary commented on src/Converter/StringToArray.php:
```diff
@@ -33,9 +33,9 @@ public function __construct($str, $delim = "")
+    public function stringToArray($limit = null)
     {
-        if (is_null($this->delim)) {
+        if (empty($this->delim)) {
```
If string `"0"` or int `0` is used as a delimiter will this still work as intended? `empty(0)` and `empty("0")` return true, which would cause no delimiter to be used.

ericpoe replied:
Thanks! Updated. Now with more tests!

johnkary commented:
Looks good aside from the two comments I made!

ericpoe committed:
Ensure toArray() returns an array

toArray() in HString wasn't getting called in the tests,
so I had no way of verifying that it was running.

toArray() in HArray was sometimes returning an \ArrayObject
instead of a standard array. This has been rectified.

ericpoe committed:
Simplify HString->toMap() when no variadic

ericpoe committed:
Update changelog
